frappe/migrate.py

```diff
@@ -110,3 +106,16 @@ def migrate(verbose=True, skip_failing=False, skip_search_index=False):
 		with open(touched_tables_file, "w") as f:
 			json.dump(list(frappe.flags.touched_tables), f, sort_keys=True, indent=4)
 		frappe.flags.touched_tables.clear()
+
+
+def run_before_migrate_hooks():
+	for app in frappe.get_installed_apps():
+		frappe.db.begin()
+		try:
+			for fn in frappe.get_hooks("before_migrate", app_name=app):
+				frappe.get_attr(fn)()
+		except Exception:
+			frappe.db.rollback()
+			raise
+		else:
+			frappe.db.commit()
```

Comment:
this change is a bit concerning

if people have added commit and/or rollback statements in their before_migrate hooks then this transaction handling might not be worth it.

also, should we handle the transaction at app level or for every hook?
Currently it's at app level i.e if any before_migrate hook failed the whole app's before_migrate hooks will *try* to rollback - didn't make much sense to me to handle it for every before_migrate hook. Though we do it on a patch level when running patches so maybe we should do this at hook level as well?